### DISCUSSION TITLE
Empty areas appearing in the dielectric geometry distribution

### DISCUSSION THREAD

**Discussion by yugeniom** [NanoComp/meep#2318] [Q&A]
Hello there,
I am experiencing issues with the buildup of the geometry of a simple dielectric stack, likely appearing our of some numerical buzz:
for certain thicknesses of the layers, I get a fictitious and thin empty layer added in the middle of the stack:

![image](https://user-images.githubusercontent.com/58916671/201372495-a0816064-02d1-4616-9400-470f5c0ac72b.png)

As I change one thickness by even less than a mesh point (here I change the upper thickness from 0.12 to 0.1195 um, with the fdtd grid resolution set to 2.5nm) , the  empty layer disappears, as in this image:

![image](https://user-images.githubusercontent.com/58916671/201372877-c5555c3f-8bf4-4667-96d3-1b354572247d.png)

The problem appears to be independent of the "eps_averaging=True/False" in the sim object setting; do you have any clue where this could come from? I add here part of the code:
>
        ALGAAS02 = mp.Medium(index=3.42241)
        ALGAAS04 = mp.Medium(index=3.30241)
        ALGAAS08 = mp.Medium(index=3.06241)
        GAAS = mp.Medium(index=3.5457)
        #ALAS = mp.Medium(index=2.9)
        AIR = mp.Medium(index=1)
        
        thk_clad = npa.float64(Sy/2)
        thk_barr_bot02 = npa.float64(0.01)
        thk_barr_bot04 = npa.float64(0.1195)
        thk_qw = npa.float64(0.02)
        thk_barr_top04 = npa.float64(0.12)
        thk_barr_top02 = npa.float64(0.02)
        thk_air = npa.float64(Sy - thk_clad - thk_barr_bot02 - thk_barr_bot04 - thk_qw - thk_barr_top04 - thk_barr_top02)

........................

        self.geometry = [
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad/2), material=ALGAAS08, size=mp.Vector3(Sx, thk_clad)
            ),  # clad
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad + thk_barr_bot02/2), material=ALGAAS02, size=mp.Vector3(Sx,thk_barr_bot02)
            ),  # bottom barrier04
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad + thk_barr_bot02 + thk_barr_bot04/2), material=ALGAAS04, size=mp.Vector3(Sx, thk_barr_bot04)
            ),  # bottom barrier02
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad + thk_barr_bot02 + thk_barr_bot04 + thk_qw/2), material=GAAS, size=mp.Vector3(Sx,thk_qw)
            ),  # qw
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad + thk_barr_bot02 + thk_barr_bot04 + thk_qw + thk_barr_top04/2), material=ALGAAS04, size=mp.Vector3(Sx,thk_barr_top04)
            ),  # top barrier04
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad + thk_barr_bot02 + thk_barr_bot04 + thk_qw + thk_barr_top04 +thk_barr_top02/2), material=ALGAAS02, size=mp.Vector3(Sx,thk_barr_top02)
            mp.Block(
                center=mp.Vector3(y=-Sy/2 + thk_clad + thk_barr_bot02 + thk_barr_bot04 + thk_qw + thk_barr_top04 +thk_barr_top02+ thk_air/2), material=AIR, size=mp.Vector3(Sx,thk_air)
            ),
        ]

>

**Comment by smartalecH** [comment]
Keep in mind that "visualizing" the permittivity is inherently difficult because it's a tensor, and different tensor components are stored at different points in space. So there's no clear solution to visualizing the permittivity at a singular point. 

So meep established a "convention", where we look at the trace of the tensor at each voxel center, interpolating when necessary. This convention seems as good as any other. 

While artifacts (like what you are seeing) sometimes appear at coarse resolutions, they will disappear in the limit of infinite resolution.

In order to verify your geometry, it's advised to crank the resolution sufficiently high to eliminate these artifacts. You can do this within plot2D itself (without having to initialize the full structure at the high resolution).

**Comment by yugeniom** [comment]
> Keep in mind that "visualizing" the permittivity is inherently difficult because it's a tensor, and different tensor components are stored at different points in space. So there's no clear solution to visualizing the permittivity at a singular point.
> 
> So meep established a "convention", where we look at the trace of the tensor at each voxel center, interpolating when necessary. This convention seems as good as any other.
> 
> While artifacts (like what you are seeing) sometimes appear at coarse resolutions, they will disappear in the limit of infinite resolution.
> 
> In order to verify your geometry, it's advised to crank the resolution sufficiently high to eliminate these artifacts. You can do this within plot2D itself (without having to initialize the full structure at the high resolution).

The problem is that I am already at a very high resolution here - with respect even to the smallest thickness in the game -, and the behaviour doesnt even depend on the subpixel smoothing, that's what puzzles me. Here  I am setting resolution = 400 , which is just 2.5nm. 
I just didnt probably get one important point from your answer: is the artifact not  just something in the visualization, isn't it? Or is it instead something just in the visualization, with no effect on the simulation results for the fields?
Here the images were retrieved from mp.Dielectric at intermediate time steps during the simulation.

**Comment by yugeniom** [comment]
Oh ok, I tryied going up to a resolution=1000; in fact the pathologic layer shrinks down as you say;
neverthless, since such a resolution is well beyond a treatable one for my cell area, can I go back to a lower resolution with no effects on the results for the fields, correct? This is just an artifact in how you reduce the information of a tensor into a scalar map for the purpose of visualization, correct?

**Comment by smartalecH** [comment]
The best way to answer questions like this is by doing a convergence analysis. Continue ramping up the resolution until the response converges.
